File: week_4/광고 삽입.py
HOUR = 3600
MINUTE = 60

def string_to_second(string):
    hh, mm, ss = string.split(':')
    return int(hh) * HOUR + int(mm) * MINUTE + int(ss)

def second_to_string(seconds):
    hh = seconds // HOUR
    seconds -= hh * HOUR
    mm = seconds // MINUTE
    seconds -= mm * MINUTE
    return ':'.join(map(lambda x: str(x).zfill(2), [hh, mm, seconds]))

def solution(play_time, adv_time, logs):
    timeline = [0] * (100 * HOUR)
    
    for log in logs:
        start, end = map(string_to_second, log.split('-'))
        timeline[start] += 1
        timeline[end] += -1
    
    adv, end = map(string_to_second, [adv_time, play_time])
    before = timeline[0]
    for i in range(1, end):
        timeline[i] += before
        before = timeline[i]
        timeline[i] += timeline[i - 1]
    
    result = 0
    maximum = timeline[adv]
    for i in range(adv + 1, end):
        total = timeline[i] - timeline[i - adv]
        if total > maximum:
            result = i - adv + 1
            maximum = total
        
    return second_to_string(result)

File: week_4/광고 삽입.js
const HOUR = 3600;
const MINUTE = 60;

String.prototype.toSecond = function() {
    const [hh, mm, ss] = this.split(':').map((str) => parseInt(str));
    if((hh && mm && ss) === undefined) return 0;
    
    return hh * HOUR + mm * MINUTE + ss;
}

Number.prototype.toTime = function() {
    let time = this;
    const hh = Math.floor(time / HOUR);
    time -= hh * HOUR;
    const mm = Math.floor(time / MINUTE);
    time -= mm * MINUTE
    return [hh, mm, time]
        .map((e) => e.toString().padStart(2, '0'))
        .join(':');
}

function solution(playTime, advTime, logs) {
    const [end, adv] = [playTime, advTime].map((str) => str.toSecond());
    const timeline = Array.from(Array(end + 1), () => 0);

    logs.forEach((log) => {
        const [start, end] = log.split('-').map((str) => str.toSecond());
        timeline[start] += 1;
        timeline[end] -= 1;
    });
    
    for(let i = 1, prev = timeline[0]; i < end; i++) {
        timeline[i] += prev;
        prev = timeline[i];
        timeline[i] += timeline[i - 1];
    }
    
    let result = 0;
    let maximum = timeline[adv];
    for(let i = adv + 1; i < end; i++){
        const total = timeline[i] - timeline[i - adv];
        if(total > maximum) {
            result = i - adv + 1;
            maximum = total;
        }
    }    
    
    return result.toTime();
}